### PR TITLE
feat: re-add external calls

### PR DESCRIPTION
**Pull request by iherger**
### Detailed summary
- Re-added external contract read functions
- Added new export functions for various modules
- Updated function names and arguments in multiple files
- Added new functions for reading contract data
- Added new functions for calculating asset values
- Added new functions for retrieving policy and fee managers
- Added new functions for retrieving external positions and vault information

> The following files were skipped due to too many changes: `packages/sdk/src/internal/Extensions/ExternalPositions/MapleLiquidity.ts`, `packages/sdk/src/internal/Extensions/IntegrationAdapters/BalancerV2Liquidity.ts`, `packages/sdk/src/internal/Extensions/Policies.ts`, `packages/sdk/src/internal/Extensions/ExternalPositions.ts`, `packages/sdk/src/Deposit.ts`, `packages/api/src/generated/enzyme/enzyme/v1/enzyme_connect.ts`, `packages/sdk/src/internal/VoteDelegation.ts`, `packages/api/src/generated/enzyme/enzyme/v1/asset_amount_pb.ts`, `packages/sdk/src/internal/Extensions/ExternalPositions/Liquity.ts`, `packages/api/src/generated/enzyme/enzyme/v1/asset_value_pb.ts`, `packages/sdk/src/internal/Extensions/Policies/CumulativeSlippageTolerance.ts`, `packages/sdk/src/internal/Extensions/IntegrationAdapters/CurveExchange.ts`, `packages/sdk/src/internal/Extensions/ExternalPositions/ConvexVoting.ts`, `packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts`, `packages/sdk/src/internal/Extensions/IntegrationAdapters/IdleV4.ts`, `packages/api/src/generated/enzyme/enzyme/v1/enzyme_pb.ts`, `packages/api/src/generated/enzyme/enzyme/v1/enums_pb.ts`, `packages/sdk/src/internal/Extensions/IntegrationAdapters/UniswapV2Liquidity.ts`, `packages/sdk/src/internal/Extensions/IntegrationAdapters/CurveLiquidity.ts`, `packages/api/src/generated/enzyme/enzyme/v1/external_position_item_pb.ts`, `packages/api/src/generated/enzyme/enzyme/v1/vault_activities_pb.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/sdk/src/Policies.ts [hunk 1]
```diff
@@ -8,13 +9,15 @@ export async function getEnabledPolicies(
   client: PublicClient,
   args: Viem.ContractCallParameters<{
     comptrollerProxy: Address;
-    policyManager: Address;
+    policyManager?: Address;
   }>,
 ) {
+  const policyManager = args.policyManager ?? (await Comptroller.getPolicyManager(client, args));
```
Comment (fubhy) [inline packages/sdk/src/Policies.ts:15]: Make the `PolicyManager` required

File: packages/sdk/src/internal/Extensions/ExternalPositions/Liquity.ts [hunk 2]
```diff
@@ -250,3 +251,104 @@ export function repayBorrowDecode(encoded: Hex): RepayBorrowArgs {
...
+
+export type LiquityTrove = {
+  debt: bigint;
+  collateral: bigint;
+  stake: bigint;
+  status: number;
+  arrayIndex: bigint;
+};
+
+export async function getLiquityTrove(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/ExternalPositions/Liquity.ts:307]: Remove `Trove`

File: packages/sdk/src/internal/Extensions/Fees.ts [hunk 3]
```diff
@@ -1,4 +1,23 @@
+import * as Abis from "@enzymefinance/abis";
+import type { Address, PublicClient } from "viem";
+import { Viem } from "../../Utils";
+
 export * as Entrance from "@enzymefinance/sdk/internal/Extensions/Fees/Entrance";
 export * as Exit from "@enzymefinance/sdk/internal/Extensions/Fees/Exit";
 export * as Management from "@enzymefinance/sdk/internal/Extensions/Fees/Management";
 export * as Performance from "@enzymefinance/sdk/internal/Extensions/Fees/Performance";
+
+export function getFeeRecipient(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/Fees.ts:10]: `getRecipient`

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/BalancerV2Liquidity.ts [hunk 4]
```diff
@@ -312,3 +312,34 @@ export function takeOrderDecode(encoded: Hex): TakeOrderArgs {
...
+
+const mintAbi = {
+  inputs: [{ internalType: "address", name: "gauge", type: "address" }],
+  name: "mint",
+  outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
+  stateMutability: "nonpayable",
+  type: "function",
+} as const;
+
+export async function getBalancerMinterRewards(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/BalancerV2Liquidity.ts:328]: getMinterRewards

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/CompoundV2.ts [hunk 5]
```diff
@@ -80,3 +81,30 @@ export function redeemDecode(encoded: Hex): RedeemArgs {
...
+  constant: true,
+  inputs: [],
+  name: "exchangeRateStored",
+  outputs: [{ name: "", type: "uint256" }],
+  payable: false,
+  stateMutability: "view",
+  type: "function",
+} as const;
+
+export function getCerc20ExchangeRateStored(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/CompoundV2.ts:99]: `getExchangeRateStore`

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts [hunk 6]
```diff
@@ -242,3 +250,95 @@ export function unstakeAndRedeemDecode(encoded: Hex): UnstakeAndRedeemArgs {
...
+) {
+  return Viem.readContract(client, args, {
+    abi: [convertCrvToCvxAbi],
+    address: args.cvxMining,
+    functionName: "ConvertCrvToCvx",
+    args: [args.amount],
+  });
+}
+
+export async function getConvexWrapperEstimateRewards(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts:281]: `getEstimatedRewards`

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts [hunk 7]
```diff
@@ -242,3 +250,95 @@ export function unstakeAndRedeemDecode(encoded: Hex): UnstakeAndRedeemArgs {
...
+
+    return tokenRewards;
+  } catch (error) {
+    // TODO: More selectively catch this error here.
+    if (error instanceof ContractFunctionExecutionError) {
+      return undefined;
+    }
+
+    throw error;
+  }
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts:316]: Remove try/catch

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts [hunk 8]
```diff
@@ -242,3 +250,95 @@ export function unstakeAndRedeemDecode(encoded: Hex): UnstakeAndRedeemArgs {
...
+}
+
+export async function getAllConvexWrapperEstimateRewards(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts:319]: `getAllEstimatedRewards`

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/CurveLiquidity.ts [hunk 9]
```diff
@@ -371,3 +381,206 @@ export function unstakeAndRedeemDecode(encoded: Hex): UnstakeAndRedeemArgs {
...
+    { type: "uint256", name: "_token_amount" },
+    { type: "int128", name: "i" },
+  ],
+  constant: true,
+  payable: false,
+  type: "function",
+  stateMutability: "view",
+} as const;
+
+export async function doesCurvePoolAllowSingleAssetRedemption(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/CurveLiquidity.ts:402]: `isSingleAssetRedemptionAllowed`

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/UniswapV2Liquidity.ts [hunk 10]
```diff
@@ -105,3 +115,191 @@ export function redeemDecode(encoded: Hex): RedeemArgs {
...
+    ],
+    name: "getPair",
+    outputs: [{ internalType: "address", name: "", type: "address" }],
+    payable: false,
+    stateMutability: "view",
+    type: "function",
+  },
+] as const;
+
+export async function getUniswapV2LendRate(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/IntegrationAdapters/UniswapV2Liquidity.ts:176]: `getLendRate`

File: packages/sdk/src/internal/Extensions/Policies.ts [hunk 11]
```diff
@@ -1,4 +1,56 @@
...
+      internalType: "uint256[]",
+      name: "listIds_",
+      type: "uint256[]",
+    },
+  ],
+  stateMutability: "view",
+  type: "function",
+} as const;
+
+export function getListIdsForPolicy(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/Policies.ts:30]: `getListIds`

File: packages/sdk/src/internal/Extensions/Policies/CumulativeSlippageTolerance.ts [hunk 12]
```diff
@@ -36,3 +39,71 @@ export function decodeSettings(settings: Hex): Settings {
     slippageTolerance,
   };
 }
+
+//--------------------------------------------------------------------------------------------
+// READ
+//--------------------------------------------------------------------------------------------
+
+export function getInfoForFund(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/Policies/CumulativeSlippageTolerance.ts:47]: `getInfo`

File: packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts [hunk 13]
```diff
@@ -59,3 +70,43 @@ export function decodeSettings(settings: Hex): Settings {
     maxInvestmentAmount,
   };
 }
+
+//--------------------------------------------------------------------------------------------
+// READ
+//--------------------------------------------------------------------------------------------
+
+export async function getEnabledMinMaxInvestmentPolicySettings(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts:78]: Remove this function

File: packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts [hunk 14]
```diff
@@ -59,3 +70,43 @@ export function decodeSettings(settings: Hex): Settings {
...
+  );
+
+  if (!hasMinMaxInvestmentPolicy) {
+    return null;
+  }
+
+  return getMinMaxInvestmentPolicySettings(client, args);
+}
+
+export function getMinMaxInvestmentPolicySettings(
```
Comment (fubhy) [inline packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts:99]: `getSettings`

File: packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import * as Abis from "@enzymefinance/abis";
+import type { Address, PublicClient } from "viem";
+import { Viem } from "../../Utils";
+
+export function getSharesWrapperDepositQueueUser(
```
Comment (fubhy) [inline packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts:5]: `getUserRequest`

File: packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+) {
+  return Viem.readContract(client, args, {
+    abi: Abis.IGatedRedemptionQueueSharesWrapperLib,
+    functionName: "getDepositQueueUserRequest",
+    address: args.sharesWrapper,
+    args: [args.depositAsset, args.user],
+  });
+}
+
+export async function getSharesWrapperRedemptionQueueUsers(
```
Comment (fubhy) [inline packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts:21]: `getUsers`

File: packages/sdk/src/internal/Tools/SharesSplitter.ts [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import * as Abis from "@enzymefinance/abis";
+import type { Address, PublicClient } from "viem";
+import { Viem } from "../../Utils";
+
+export function getSharesSplitterTokenBalanceClaimableForUser(
```
Comment (fubhy) [inline packages/sdk/src/internal/Tools/SharesSplitter.ts:5]: `getClaimableTokenBalance`

File: packages/sdk/src/Deposit.ts [hunk 18]
```diff
@@ -134,3 +135,43 @@ export async function isAllowedDepositor(
...
+
+    return output;
+  } catch (error) {
+    // TODO: More selectively catch this error here.
+    if (error instanceof ContractFunctionExecutionError) {
+      return undefined;
+    }
+
+    throw error;
+  }
```
[the remark this answers is not in the document]
Comment (iherger) [reply packages/sdk/src/Deposit.ts]: Fixed.

File: packages/sdk/src/Fees.ts [hunk 19]
```diff
@@ -70,3 +72,48 @@ export function doesAutoProtocolFeeSharesBuyback(
...
+    comptrollerProxy?: Address;
+  }>,
+) {
+  const [sharesSupply, _comptrollerProxy] = await Promise.all([
+    Viem.readContract(client, args, {
+      abi: Abis.IVaultLib,
+      functionName: "totalSupply",
+      address: args.vaultProxy,
+    }),
+    ...(args.comptrollerProxy === undefined ? [getComptrollerProxy(client, { vaultProxy: args.vaultProxy })] : []),
```
[the remark this answers is not in the document]
Comment (iherger) [reply packages/sdk/src/Fees.ts]: Done.

File: packages/sdk/src/Policies.ts (hunk 1 above)
Comment (iherger) [reply packages/sdk/src/Policies.ts]: Done.

File: packages/sdk/src/internal/Extensions/ExternalPositions/ConvexVoting.ts [hunk 20]
```diff
@@ -174,3 +175,116 @@ export function delegateDecode(encoded: Hex): DelegateArgs {
...
+    total,
+    unlockable,
+    locked,
+    lockedData,
+  };
+
+  return lockedBalancesData;
+}
+
+export async function getAllVoteLockedConvexTokenLockedBalances(
```
[the remark this answers is not in the document]
Comment (iherger) [reply packages/sdk/src/internal/Extensions/ExternalPositions/ConvexVoting.ts]: Done.

File: packages/sdk/src/internal/Extensions/ExternalPositions/Liquity.ts (hunk 2 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/ExternalPositions/Liquity.ts]: Actually removed `Liquity`

File: packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts (hunk 13 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts]: Removed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/BalancerV2Liquidity.ts (hunk 4 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/BalancerV2Liquidity.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/CompoundV2.ts (hunk 5 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/CompoundV2.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts (hunk 6 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts (hunk 7 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts]: Removed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts (hunk 8 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/ConvexCurveLpStaking.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/CurveLiquidity.ts (hunk 9 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/CurveLiquidity.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/IntegrationAdapters/UniswapV2Liquidity.ts (hunk 10 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/IntegrationAdapters/UniswapV2Liquidity.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/Policies.ts (hunk 11 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/Policies.ts]: Renamed.

File: packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts (hunk 14 above)
Comment (iherger) [reply packages/sdk/src/internal/Extensions/Policies/MinMaxInvestment.ts]: Renamed.

File: packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts (hunk 15 above)
Comment (iherger) [reply packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts]: Renamed to `getDepositQueueUserRequest`

File: packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts (hunk 16 above)
Comment (iherger) [reply packages/sdk/src/internal/Tools/GatedRedemptionQueueSharesWrapper.ts]: Renamed to `getRedemptionQueueUsers`

File: packages/sdk/src/internal/Tools/SharesSplitter.ts (hunk 17 above)
Comment (iherger) [reply packages/sdk/src/internal/Tools/SharesSplitter.ts]: Renamed.

**Pull request merged**